hadoop-hdfs-project/hadoop-hdfs-native-client/src/main/native/libhdfs/hdfs.c

```diff
@@ -2512,17 +2509,35 @@ int hadoopRzOptionsSetByteBufferPool(
       if (jthr) {
           printExceptionAndFree(env, jthr, PRINT_EXC_ALL,
               "hadoopRzOptionsSetByteBufferPool(className=%s): ", className);
-          errno = EINVAL;
-          return -1;
+          ret = EINVAL;
+          goto done;
       }
-      opts->byteBufferPool = (*env)->NewGlobalRef(env, byteBufferPool);
-      if (!opts->byteBufferPool) {
-          printExceptionAndFree(env, jthr, PRINT_EXC_ALL,
+      // Only set opts->byteBufferPool if creating a global reference is
+      // successful
+      globalByteBufferPool = (*env)->NewGlobalRef(env, byteBufferPool);
+      if (!globalByteBufferPool) {
+          printPendingExceptionAndFree(env, PRINT_EXC_ALL,
                   "hadoopRzOptionsSetByteBufferPool(className=%s): ",
                   className);
-          errno = EINVAL;
-          return -1;
+          ret = EINVAL;
+          goto done;
+      }
+      // Delete any previous ByteBufferPool we had before setting a new one.
+      if (opts->byteBufferPool) {
+          (*env)->DeleteGlobalRef(env, opts->byteBufferPool);
       }
+      opts->byteBufferPool = globalByteBufferPool;
+    } else if (opts->byteBufferPool) {
+        // If the specified className is NULL, delete any previous
+        // ByteBufferPool we had.
+        (*env)->DeleteGlobalRef(env, opts->byteBufferPool);
```

Comment:
need to also set it back to null, right?

Reply:
Yeah done.